conjure-runtime/src/service/node/limiter/deficit_semaphore.rs

```diff
@@ -0,0 +1,573 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use parking_lot::Mutex;
+use pin_project::{pin_project, pinned_drop};
+use std::future::Future;
+use std::marker::PhantomPinned;
+use std::pin::Pin;
+use std::ptr::NonNull;
+use std::sync::Arc;
+use std::task::{Context, Poll, Waker};
+
+/// An async semaphore with the ability to arbitrarily adjust its total number of permits. It is absolutely fair - i.e.
+/// tasks will be granted permits in exactly the order they began attempting to acquire them.
+///
+/// A traditional semaphore only tracks the current number of available permits, rather than the "total" number of
+/// permits in the system, including those currently checked out by tasks. This semaphore instead tracks both that total
+/// as well as the number of permits checked out. The total permit count can be adjusted arbitrarily, even if it would
+/// cause the available permit count to go negative.
+///
+/// It's implemented with an intrusive linked list of waiters, taking advantage of the fact that futures are pinned in
+/// memory.
+pub struct DeficitSemaphore(Mutex<State>);
+
+impl DeficitSemaphore {
+    pub fn new(permits: usize) -> Arc<DeficitSemaphore> {
+        Arc::new(DeficitSemaphore(Mutex::new(State {
+            total_permits: permits,
+            outstanding_permits: 0,
+            head: None,
+            tail: None,
+        })))
+    }
+
+    pub fn add_permits(&self, permits: usize) {
+        let mut state = self.0.lock();
+        state.total_permits = state
+            .total_permits
+            .checked_add(permits)
+            .expect("permit count overflow");
+        state.maybe_wake();
+    }
+
+    pub fn remove_permits(&self, permits: usize) {
+        let mut state = self.0.lock();
+        state.total_permits = state
+            .total_permits
+            .checked_sub(permits)
+            .expect("permit count underflow");
+    }
+
+    pub fn acquire(self: Arc<Self>) -> Acquire {
+        Acquire {
+            semaphore: self,
+            node: None,
+            _p: PhantomPinned,
+        }
+    }
+}
+
+struct State {
+    total_permits: usize,
+    outstanding_permits: usize,
+    head: Option<NonNull<Node>>,
+    tail: Option<NonNull<Node>>,
+}
+
+impl State {
+    fn has_permits(&self) -> bool {
+        self.total_permits > self.outstanding_permits
+    }
+
+    fn maybe_wake(&mut self) {
+        if self.has_permits() {
+            if let Some(head) = &mut self.head {
+                unsafe {
+                    if let Some(waker) = head.as_mut().waker.take() {
+                        waker.wake();
+                    }
+                }
+            }
+        }
+    }
+}
+
+struct Node {
+    waker: Option<Waker>,
+    next: Option<NonNull<Node>>,
+    prev: Option<NonNull<Node>>,
+}
+
+impl Node {
+    // SAFETY: this node must be in `state`'s queue.
```

Comment:
Is it worth it to add tests that exercise the debug assertions?

Reply:
The debug assertions check for situations that should be impossible - the stress_test test is probably the closest thing to what you're thinking of where we just perform a bunch of random concurrent operations to see if it breaks.